Bug Fix for Node Email Notifications

ExtremeFiretop opened the pull request:
Bug Fix for Node Email Notifications
I found some bugs with the node feature with this latest firmware release btw that are in our current production version:

1. Even though the node has a name, the email says TBD

![image](https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/assets/1971404/90cd9988-2c50-4c02-aba8-b25378da32d3)

![image](https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/assets/1971404/931caa55-84d4-4a32-be6d-991263b3d5dd)

2. I'm not sure if the email title is accurate, in my case the nodes are GT-AXE11000's but I don't remember making that dynamic. Will need to review.

![image](https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/assets/1971404/665b756a-78db-4d93-93c9-44d9141594ab)

3. I am able to trigger the email every time I go to the nodes menu.

ExtremeFiretop commented:
Old version of the function was broken at some point and I didn't even notice:

![image](https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/assets/1971404/0e1d0810-8e17-45ca-bdb5-82e7ae628a55)

ExtremeFiretop commented:
New version works correctly:

![image](https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/assets/1971404/613c6724-547d-4804-805b-b6ae9e7c710e)

Martinski4GitHub commented on MerlinAU.sh:
```diff
@@ -806,15 +806,18 @@ _GetAllNodeSettings_()
+        local escaped_mac_address=$(echo "$node_mac_address" | sed 's/[\*\.]/\\&/g')
+
+        # Construct the pattern to match the entire line containing the MAC address and the setting part
+        pattern="${escaped_mac_address}.*${setting_part}=\"[^\"]*\""
```
The regex seems to be more complicated than it needs to be, but perhaps I'm missing something because I don't have any AiMesh nodes so I cannot really recreate the problem (I didn't fully understand the old code either because of the same issue: no mesh nodes).

What's the exact key/value pattern that you're trying to find/match in the settings file? 
Can you give me a couple of examples from your settings file (just obfuscate your MAC addresses)?

I think I need more context to understand what you're trying to do here so I can wrap my head around it.

ExtremeFiretop replied:
@Martinski4GitHub 

Sorry for the delay, I had other things come up that took priority. I know you understand.
I did read the message but wasn't in a situation were I could respond at the moment.

At your request is is the contents of my settings file:

```
Node_7C:XX:XX:XX:XX:XX_Model_NameID="GT-AXE11000-DBF0"
Node_7C:XX:XX:XX:XX:XX_New_Notification_Date="2024-04-26_19:22:00"
Node_7C:XX:XX:XX:XX:XX_New_Notification_Vers="3004.388.7.0"
```

![image](https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/assets/1971404/d1113e10-a903-401e-b94e-f1880b0672ca)

ExtremeFiretop replied:
And every MAC address is the same for the one node I'm trying to look up.

ExtremeFiretop replied:
Resolved in commit: https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/pull/197/commits/48349af30f7310ef9f8cb0e71d8cf2cd7075191f

Martinski4GitHub replied:
> 
> At your request is is the contents of my settings file:
> 
> ```
> Node_7C:XX:XX:XX:XX:XX_Model_NameID="GT-AXE11000-DBF0"
> Node_7C:XX:XX:XX:XX:XX_New_Notification_Date="2024-04-26_19:22:00"
> Node_7C:XX:XX:XX:XX:XX_New_Notification_Vers="3004.388.7.0"
> ```
>
Thank you. This provides the necessary context to understand what your code is trying to do. With this in mind, now I  can review & follow the code to try to wrap my head around it.

Martinski4GitHub commented on MerlinAU.sh:
```diff
@@ -806,15 +806,18 @@ _GetAllNodeSettings_()
         if [ -n "$matched_lines" ]; then
-            # If multiple lines match, this extracts the value from the first matched line
-            setting_value="$(echo "$matched_lines" | head -n 1 | awk -F '=' '{print $2}' | tr -d '"')"
+            # Extract the value from the first matched line
```
What's the reasoning for removing the enclosing double quotes from variable assignments?

ExtremeFiretop replied:
No specific reason I think that was just due to me copy and pasting from my old submission that worked in PR: https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/pull/179/files

![image](https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/assets/1971404/6145d183-f73e-42f3-8449-4a072dd5e0d0)

I normally do that incase there was some kinda change I didn't see, I can fix it I doubt this had anything to do with the fix. Was just left over from me copy and pasting from the original function when testing stuff.

ExtremeFiretop replied:
Added them back and added the new ones from my fix

Confirmed this had no impact either way, it was just left from my original method while testing.

ExtremeFiretop commented:
Another "bug" identified here:

https://www.snbforums.com/threads/merlinau-v1-1-1-the-ultimate-firmware-auto-updater-now-available-in-amtm.88577/post-904357

ExtremeFiretop commented:
Another bug, my script as a cron job got stuck here:

```
    if "$isInteractive"
    then
        printf "${GRNct}**IMPORTANT**:${NOct}\nThe firmware flash is about to start.\n"
        printf "Press <Enter> to stop now, or type ${GRNct}Y${NOct} to continue.\n"
        printf "Once started, the flashing process CANNOT be interrupted.\n"
        if ! _WaitForYESorNO_ "Continue?"
        then
            Say "F/W Update was cancelled by user."
            _DoCleanUp_ 1 "$keepZIPfile" "$keepWfile"
            return 1
        fi
    fi
```

ExtremeFiretop commented:
@Martinski4GitHub 

Resolved the cron issue in commit: https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/pull/197/commits/1b9bbfc347fbe18850597e456c48f5e0f570aad9

ExtremeFiretop commented:
@Martinski4GitHub 

I'll give you today to review obviously, but if I don't hear by end of day I'll just push this through as-is.
This code works and the crons currently don't unfortunately it seems, I don't want too many people upset.

It's a fairly small change, and this is only a PR to go to dev lol, so if after the fact you feel that the regex is too complicated, we can simplify it, feel free to make a suggestion in the this PR or the next PR! :)

ExtremeFiretop commented:
@Martinski4GitHub 

Could it be simplified by only searching for colons instead of asterisks ?
I'll test, maybe that will satisfy your concerns if it works. :)

ExtremeFiretop commented:
I tried:

local escaped_mac_address="$(echo "$node_mac_address" | sed 's/[:]/\\&/g')"

But that only kinda worked?

![image](https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/assets/1971404/3b50aef2-c8ea-4bf8-b262-7e26ca3aff95)

I'll do more research, I've never been strong on regex.

ExtremeFiretop commented:
LOL. Okay. so grep doesn't need colons (:) to be escaped because they are not special characters in regex.

Resolved in commit: https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/pull/197/commits/48349af30f7310ef9f8cb0e71d8cf2cd7075191f

ExtremeFiretop commented:
> Another "bug" identified here:
> 
> https://www.snbforums.com/threads/merlinau-v1-1-1-the-ultimate-firmware-auto-updater-now-available-in-amtm.88577/post-904357

Resolved this bug with commit: https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/pull/197/commits/d2f86af7c833992d0593a1a51166c95c2d203150

ExtremeFiretop commented:
> Another bug, my script as a cron job got stuck here:
> 
> ```
> 2024-04-27 04:31:07 Required RAM: 227487 KB - RAM Free: 117240 KB - RAM Available: 389221 KB
> 2024-04-27 04:31:07 No high-risk phrases found in the change-log.
> 2024-04-27 04:31:07 Required RAM: 227487 KB - RAM Free: 117820 KB - RAM Available: 389689 KB
> 2024-04-27 04:31:07 ROG Build selected for flashing
> 2024-04-27 04:31:08 Required RAM: 227487 KB - RAM Free: 116660 KB - RAM Available: 388529 KB
> 2024-04-27 04:31:08 Router Web URL is: http://192.168.50.1
> ```

I just also realize there is high-risk phrases in the changelog yet for some reason the log check said there wasn't, but when I ran interactively just now testing the latest change, it did pop-up. I wonder what that is about.

ExtremeFiretop commented:
@Martinski4GitHub 

I made a new PR for that specific bug.
I'm still giving you until end of day my time. (probably around 10PM my time) before I send this all out.

But like I said I gotchu, I know your busy I can handle some patches ;)
You can clean them up afterwards if you see fit

Martinski4GitHub commented:
> Another "bug" identified here:
> 
I really hate when people send screenshots from their smartphones and believe the tiny text will somehow become readable when posted, especially given how the image was captured. I know people think they're being helpful, and often they are trying to be, but don't they realize how "unreadable" the screenshot is when posted on the forum? (sorry, it's just one of my pet peeves).

ExtremeFiretop commented:
> > Another "bug" identified here:
> 
> I really hate when people send screenshots from their smartphones and believe the tiny text will somehow become readable when posted, especially given how the image was captured. I know people think they're being helpful, and often they are trying to be, but don't they realize how "unreadable" the screenshot is when posted on the forum? (sorry, it's just one of my pet peeves).

I think I lost some of my 20/20 vision trying to make out what happened in that screenshot 😂

Martinski4GitHub commented:
> LOL. Okay. so grep doesn't need colons (:) to be escaped because they are not special characters in regex.
> 
Yes, exactly. That's one of the reasons the regex looked unnecessarily complicated as it was. It's better now but I still think it can be further simplified given the format of the node entries in the settings file. Yes, it worked as it was before and still works as it is now, but when it comes to regex coding I often like to make it simpler while still matching the required strings.

I'll review the code much closer now that I have the samples from your nodes.

Martinski4GitHub commented:
> > Another "bug" identified here:
> > https://www.snbforums.com/threads/merlinau-v1-1-1-the-ultimate-firmware-auto-updater-now-available-in-amtm.88577/post-904357
> 
> Resolved this "bug" with commit: ...
>
> 
> Again this change wasn't strictly required because the "bug" was the changelog itself. But when testing it threw me off that it was using the buildno on the firmware instead of the zip.
>
I'm not following why the change was made while not strictly necessary. IOW, how is the change better than what it was before?

The "**buildno**" NVRAM variable is used to create the version number of the current F/W release installed on the router, which is then used to compare against the next F/W release version, so the 1st part of the version string (**386** or **388**) must match, otherwise we have different F/W branches. So using the value from "**buildno**" was not a problem at all, so I'm confused as to why it was modified. Was it causing an actual problem?

ExtremeFiretop commented:
> > > Another "bug" identified here:
> > > https://www.snbforums.com/threads/merlinau-v1-1-1-the-ultimate-firmware-auto-updater-now-available-in-amtm.88577/post-904357
> > 
> > 
> > Resolved this "bug" with commit: ...
> > Again this change wasn't strictly required because the "bug" was the changelog itself. But when testing it threw me off that it was using the buildno on the firmware instead of the zip.
> 
> I'm not following why the change was made while not strictly necessary. IOW, how is the change better than what it was before?
> 
> The "**buildno**" NVRAM variable is used to create the version number of the current F/W release installed on the router, which is then used to compare against the next F/W release version, so the 1st part of the version string (**386** or **388**) must match, otherwise we have different F/W branches. So using the value from "**buildno**" was not a problem at all, so I'm confused as to why it was modified. Was it causing an actual problem?

Hi @Martinski4GitHub 

I'm up! I'm up! Hahaha Funny enough I knew this question was coming so I prepped my defense on this one last night ;)

The decision to switch to the new code, which uses an if-else structure with the $release_version variable, was mostly to maintain the original functionality while enhancing the script’s flexibility, readability, and ease of debugging. 

To my understanding this approach aligns well with common best practices in software development, especially in environments where multiple developers might interact with the code or where frequent testing and changes are necessary. please correct me here if I am wrong. 

As I tried to touch on here:

> Again this change wasn't strictly required because the "bug" was the changelog itself. But when testing it threw me off that it was using the buildno on the firmware instead of the zip.

And here:

> Just because it wasn't the behavior I expected to see when I put the script in debug mode to fake another model.

The answer I prepared is:

During my initial tests in debug mode, the original codes behavior of relying directly the "**nvram get buildno command**" proved confusing and less predictable. The goal here was not to 'make it better' in anyway, don't take it that way at all.
But to make it easier to control without modifying the with the system's NVRAM or the scripts production logic. :)

The change I made uses a variable for the "version number" which allows you to set and alter only the "ProductID" freely in a testing or debugging scenario. (debug mode note at the top of the script here:)

![image](https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/assets/1971404/9bc83005-cf0a-494e-bf51-7504b3f18ddb)

This offers a clearer and more controlled testing process (at least for me!) 

Using a variable ($release_version) as the source for the version number makes the code adaptable. If the logic needs to be extended or modified for example to add new condition,  it's straightforward within an if-else framework. Your original code was extremely compact and efficient (as all your code is!). But I think it might add a level of complexity that might not provide additional value given the task it performs. 

> The regex seems to be more complicated than it needs to be

Similar to how you mentioned the regex was more complicated than it needs to be above, and it was a concern, by simplifying the approach, the script becomes less error-prone and easier to manage over time, I think this is something we both understand well, especially in a dynamic development environment like Github where anyone can come in with a PR to help.

The original goal of the script to assign a specific tag based on the presence of "386" in the version number remains intact with the new code. It meets the requirement to be just as effective as the original, ensuring that there is no loss in functionality. 
Both versions rely on grep, but the original method also depends on nvram. This can be seen as an additional dependency that might not be strictly speaking required.

(if we can minimize the amount of nvram calls I'm all for it! Always have been since my test router started having memory issues with nvram)

So in short you are correct, this was not required for the operation of the script but was implemented to address specific issues encountered during testing and readability.  The shift to an if-else structure simplifies the understanding of the code's logic. It's a format which is universally recognized and easier for most developers to follow and maintain, compared to a one-liner with nested commands and logical operators. This clarity is beneficial not only for current maintainers but also for any new developers who might work with the script in the future once we are both are long gone. This familiarity of if-else tends to make the code easier to understand, maintain, and debug and this approach can reduce the learning curve for anyone else.

Was there a specific reason you wanted to keep it the way it was? Instead of the change I made? Maybe I am missing something, feel free to defend and explain your viewpoint on this, and we can make a final decision for the next release :)

ExtremeFiretop commented:
Btw I know your only asking for clarification to understand the reasoning behind it, I'm not trying to push any specific way.

And this time I know your askingnot because you took it personally (I understand this now between us) hahah but because your trying to align with my though process.

 just figured I'd have the discussion as I mentioned yesterday I figured you would tag me in one of the PRs about this and you'd just want to have a discussion around it.

ExtremeFiretop commented:
> > LOL. Okay. so grep doesn't need colons (:) to be escaped because they are not special characters in regex.
> 
> Yes, exactly. That's one of the reasons the regex looked unnecessarily complicated as it was. It's better now but I still think it can be further simplified given the format of the node entries in the settings file. Yes, it worked as it was before and still works as it is now, but when it comes to regex coding I often like to make it simpler while still matching the required strings.
> 
> I'll review the code much closer now that I have the samples from your nodes.

I'm all open to sending another release! No reason why not!

I was also thinking we haven't bumped the min version in a while... Just food for thought.

ExtremeFiretop commented:
Just updated my side gigs router to the latest version of MerlinAU
Then tried to run it and got this error at the end:

![image](https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/assets/1971404/c0e32a65-ed3f-4467-b9f5-d6a74e5c5d5a)

The router then went offline and came back online properly updated. So not sure what the cause of that was.

Martinski4GitHub commented:
> Btw I know your only asking for clarification to understand the reasoning behind it, I'm not trying to push any specific way.
> 
> And this time I know your asking not because you took it personally (I understand this now between us) hahah but because your trying to align with my though process.
> 
> Just figured I'd have the friendly open discussion around it. As I mentioned yesterday I figured you would tag me in one of the PRs about this and you'd just want to have a discussion around it.
> 
> I hope my answer above was sufficient. But I'm totally open to having a discussion around the change if it wasn't a sufficient explanation.
> 
> Quite frankly it comes down to me testing in the forums, re-creating his issue, and then realizing it was due to how I tested. I had one of those "I wonder why we did it this way" moments and just went about my day changing it how I thought I remembered it working when I reviewed it the first time. So I'm not personally attached to this method, it just worked better when debugging that's pretty much literally the only reason I did it at that moment.
>

Just you know where I'm coming from and my professional perspective. In s/w development, as a general rule, we have 9 broad categories for changing code:

1. To fix/address bugs.
2. To add new features/functionality.
3. To improve existing functionality.
4. To make code more robust & efficient.
5. To follow recognized best coding practices.
6. To comply with established coding standards.
7. To improve code readability.
8. To make debugging easier.
9. Due to some refactoring changes.

Some of these categories may have overlapping goals, but the idea is to have a clear intention for making code changes. So every time I do code reviews (either at work as part of my job, or here in GitHub projects), my mind automatically looks for the reason(s) to fully understand the modifications and the ultimate goal(s).

Based on your explanation, I understand that it was specifically for category **8** above, which I agree with. As you said, I was trying to understand the reasoning behind it. Most of the time I get it right away, but this time it was a little confusing due to your message that first indicated that it was to fix a bug, but then it wasn’t; it was something else but it was not entirely clear to me what the reason was.

Thanks for the clarification.

ExtremeFiretop commented:
> > Btw I know your only asking for clarification to understand the reasoning behind it, I'm not trying to push any specific way.
> > And this time I know your asking not because you took it personally (I understand this now between us) hahah but because your trying to align with my though process.
> > Just figured I'd have the friendly open discussion around it. As I mentioned yesterday I figured you would tag me in one of the PRs about this and you'd just want to have a discussion around it.
> > I hope my answer above was sufficient. But I'm totally open to having a discussion around the change if it wasn't a sufficient explanation.
> > Quite frankly it comes down to me testing in the forums, re-creating his issue, and then realizing it was due to how I tested. I had one of those "I wonder why we did it this way" moments and just went about my day changing it how I thought I remembered it working when I reviewed it the first time. So I'm not personally attached to this method, it just worked better when debugging that's pretty much literally the only reason I did it at that moment.
> 
> Just you know where I'm coming from and my professional perspective. In s/w development, as a general rule, we have 9 broad categories for changing code:
> 
> 1. To fix/address bugs.
> 2. To add new features/functionality.
> 3. To improve existing functionality.
> 4. To make code more robust & efficient.
> 5. To follow recognized best coding practices.
> 6. To comply with established coding standards.
> 7. To improve code readability.
> 8. To make debugging easier.
> 9. Due to some refactoring changes.
> 
> Some of these categories may have overlapping goals, but the idea is to have a clear intention for making code changes. So every time I do code reviews (either at work as part of my job, or here in GitHub projects), my mind automatically looks for the reason(s) to fully understand the modifications and the ultimate goal(s).
> 
> Based on your explanation, I understand that it was specifically for category **8** above, which I agree with. As you said, I was trying to understand the reasoning behind it. Most of the time I get it right away, but this time it was a little confusing due to your message that first indicated that it was to fix a bug, but then it wasn’t; it was something else but it was not entirely clear to me what the reason was.
> 
> Thanks for the clarification.

No worries! I just hope I got my reasoning across okay, I'm totally cool with chatting about these types of changes in the future as well if you have anymore "why did he change this"

Like the double quotes, great catch, I honestly truthfully didn't even notice. And that's just because of the language I was originally trained on. In powershell I never use the outside double quotes so my brain just.. idk it skips that lol.

I didn't even register they were missing until you said something and I was like.. yeah I totally rolled back the cleanup in that function for no reason at all. So I rolled it back.

But I can promise thar change I had a reason haha! But I agree me calling it a "bug" wasn't clear.

I was trying to use the quotes to make it seem like the bug wasn't really a "bug" lol 😂 But really due to the missing changelog and me failing to identify it using the buildno before me flipping the switch.

Martinski4GitHub commented:
> > > LOL. Okay. so grep doesn't need colons (:) to be escaped because they are not special characters in regex.
> > 
> > 
> > Yes, exactly. That's one of the reasons the regex looked unnecessarily complicated as it was. It's better now but I still think it can be further simplified given the format of the node entries in the settings file. Yes, it worked as it was before and still works as it is now, but when it comes to regex coding I often like to make it simpler while still matching the required strings.
> > I'll review the code much closer now that I have the samples from your nodes.
> 
> I'm all open to sending another release! No reason why not!
> 
> I was also thinking we haven't bumped the min version in a while... Just food for thought.
>
Currently, I see no reason to issue another release soon after the last one. IOW, there's no urgency to fix something, as far as I can tell, so we're good.

Martinski4GitHub commented:
> 
> During my initial tests in debug mode, the original codes behavior of relying directly the "**nvram get buildno**" command proved confusing and less predictable. (As found in this forums thread):
> ...
>
This part is still puzzling & vague to me. Are you saying that the "**buildno**" NVRAM value is not reliable or predictable anymore? If so, under what circumstances is this the case? Can you elaborate? Perhaps there's some miscommunication but I like to make sure.

ExtremeFiretop commented:
> > During my initial tests in debug mode, the original codes behavior of relying directly the "**nvram get buildno**" command proved confusing and less predictable. (As found in this forums thread):
> > ...
> 
> This part is still puzzling & vague to me. Are you saying that the "**buildno**" NVRAM value is not reliable or predictable anymore? If so, under what circumstances is this the case? Can you elaborate? Perhaps there's some miscommunication but I like to make sure.

Ahhhh, I see the confusion. No the buildno is perfectly reliable and predictable in its returned results. Not what I meant, what I was trying to convey was that it wasn't making the function act in a predictable manner when faking a different model.

When I changed that ProductID value, I expected the entire update function of the script to now function like it was that model, the fact that 90% of the script did but then didn't at the changelog step, is what I meant but I wasnt as predictable. When I say predictable I meant the entirety of the update function, not the returned results of that nvram value, the returned results are absolutely predictable, but the update function not running in its entirety when I faked a model is what was unpredictable to me, at least without me having to modify the production logic in that area as well.

But because I was specific trying to test that part of the production code I didn't want to change anything in the code itself I wanted to change the information being fed into that part of the code. Does that make more sense?

Martinski4GitHub commented:
> 
> Ahhhh, I see the confusion. No the buildno is perfectly reliable and predictable in its returned results. Not what I meant, what I was trying to convey was that it wasn't making the function act in a predictable manner when faking a different model.
> 
> When I changed that ProductID value, I expected the entire update function of the script to now function like it was that model, the fact that 90% of the script did but then didn't at the changelog step, is what I meant by It wasnt as predictable. When I say predictable I meant the entirety of the update function, not the returned results of that nvram value, the returned results are absolutely predictable, but the update function not running in its entirety when I faked a model is what was unpredictable to me, at least without me having to modify the production logic in that area as well.
> 
> But because I was specific trying to test that part of the production code I didn't want to change anything in the code itself I wanted to change the information being fed into that part of the code. Does that make more sense?
> 
> Edit: I'm on my phone the amount of typos right now is absurd.
>
Ah, OK. I'm starting to see more clearly now your reasoning. Basically, your testing/debugging methodology was giving you "bad" or "unpredictable" results, but the "**buildno**" value itself was OK and as expected based on the the F/W that was currently installed. I think I got it now.

ExtremeFiretop commented:
> > Ahhhh, I see the confusion. No the buildno is perfectly reliable and predictable in its returned results. Not what I meant, what I was trying to convey was that it wasn't making the function act in a predictable manner when faking a different model.
> > When I changed that ProductID value, I expected the entire update function of the script to now function like it was that model, the fact that 90% of the script did but then didn't at the changelog step, is what I meant by It wasnt as predictable. When I say predictable I meant the entirety of the update function, not the returned results of that nvram value, the returned results are absolutely predictable, but the update function not running in its entirety when I faked a model is what was unpredictable to me, at least without me having to modify the production logic in that area as well.
> > But because I was specific trying to test that part of the production code I didn't want to change anything in the code itself I wanted to change the information being fed into that part of the code. Does that make more sense?
> > Edit: I'm on my phone the amount of typos right now is absurd.
> 
> Ah, OK. I'm starting to see more clearly now your reasoning. Basically, your testing/debugging methodology was giving you "bad" or "unpredictable" results, but the "**buildno**" value itself was OK and as expected based on the the F/W that was currently installed. I think I got it now.

Bingo! You got it, this is the only reason I changed it. Now all the defense I wrote up I think still holds up, but that's really the frank  honest answer lol.

If you were to roll back the the old version of the script, and change that ProductID value, you would see the script would run all the way to the end where it prompts to confirm the flash, downloads that models firmwares etc. but for "some reason" (I know now lol) that changelog step wasn't functional when faking a model.

Now normally, I'd just change that code and rerun, but because that part of the code was what was being reported as problematic and I didnt want to change the code I just wanted to see it "run as his model"

I had not even looked at the code yet to identify it was using buildno, I just ran it as his model, saw the issue and freaked out.

The truth why I saw his issue is because the script downloaded his firmware, then it checked my buildno, saw it was 388 instead of 386, looked for the changelog-NG file which wasn't there anymore because RMerlin had fixed it already, and spit that error out.

Now in his case the reason was because the script did look for 386 and changelo-386.txt but he had changelog-NG in his zip file. But I didn't know that yet, I was just starting to test, I just wanted to see "what his model did" and the first run if you read the thread you'll see I said I was unable to replicate his issue at first. Lol. Because RMerlin had not fixed the zip yet, so when I ran in faking the model, it looked at my buildno, saw 388, looked for changelog-NG, found it in his firmware zip, all worked, when it shouldn't have. 

Then i realized I'd have to change that code to properly test it.

I did, the code was functional, there was nothing wrong with it, other than it not running how I predicted in a "debug mode" faking another model. So that's basically the jist of it all.

ExtremeFiretop commented:
@Martinski4GitHub 

I'm actually misremembering, because I wrote up that I was unable to replicate his issue, but looking back I didn't post it. Instead I posted that it was due to the changelog being missing, but I had only identified that cause I had manually downloaded his zip and looked inside.

Martinski4GitHub commented:
> > 
> > 
> > Ah, OK. I'm starting to see more clearly now your reasoning. Basically, your testing/debugging methodology was giving you "bad" or "unpredictable" results, but the "**buildno**" value itself was OK and as expected based on the the F/W that was currently installed. I think I got it now.
> 
> Bingo! You got it, this is the only reason I changed it. Now all the defense I wrote up I think still holds up, but that's really the frank honest answer lol.
> 
Well, a couple of points are debatable & extremely subjective so those would not hold up, IMO (e.g. level of "complexity" of a ternary expression vs. if-else conditional); but I really don't want to debate those points right now. It's good enough for me that it was for easier debugging, and ultimately it's a good reason all by itself.

ExtremeFiretop commented:
> > > Ah, OK. I'm starting to see more clearly now your reasoning. Basically, your testing/debugging methodology was giving you "bad" or "unpredictable" results, but the "**buildno**" value itself was OK and as expected based on the the F/W that was currently installed. I think I got it now.
> > 
> > 
> > Bingo! You got it, this is the only reason I changed it. Now all the defense I wrote up I think still holds up, but that's really the frank honest answer lol.
> 
> Well, a couple of points are debatable & extremely subjective so those would not hold up, IMO (e.g. level of "complexity" of a ternary expression vs. if-else conditional); but I really don't want to debate those points right now. It's good enough for me that it was for easier debugging, and ultimately it's a good reason all by itself.

Fair enough haha, I didn't mean it was complex to read, just compared to an if else statement it's more condensed and efficient, but not as "user friendly" to read.

But I like to think that someday even if we aren't interested in this project anymore, someone else might want to keep it alive and take it on. (Not happening anytime soon on my watch! Hahah) But jokes aside if that happens i don't know what their level will be, so I feel making everything longer and less condensed but more user readable isn't a bad exchange. :)

That's all I was trying to say, it's not complex it's just required a higher level of understanding, as I touched on, you may need to be higher up the learning curve to be able to just look at it and know what it means, compared to someone just learning and seeing if and else statements.

I myself think it can become a bit long and gross and unnecessary at times as well. But I also think it's very human readable. I guess that's the only thing I was trying to point out why I default to that for this project when possible.
